runtime/doc/lsp.txt

```diff
@@ -1323,14 +1323,75 @@ save({lenses}, {bufnr}, {client_id})                 *vim.lsp.codelens.save()*
 ==============================================================================
 Lua module: vim.lsp.semantic_tokens                      *lsp-semantic_tokens*
 
+add_rule({rule}, {opts})                  *vim.lsp.semantic_tokens.add_rule()*
+    Add a new semantic token highlighting rule.
+
+    Example: >lua
+    -- Highlight static local variables
+    vim.lsp.semantic_tokens.add_rule(function(token)
+      if token.modifiers.functionScope and token.modifiers.static then
+        return "MyStaticLocalHL"
+      end
+    end)
+
+    vim.api.nvim_set_hl(0, "MyStaticLocalHL", { fg = "red" })
+<
+
+    Parameters: ~
+      • {rule}  (function) Lua callback invoked on visible tokens. Return a
+                highlight group name to apply that highlight to a token. Args:
+                • token: a table with the fields:
+                  • type: string LSP token type, e.g. 'function' or 'variable'
+                  • modifiers: set of LSP token modifiers, e.g. >lua
+
+                     { defaultLibrary = true, readOnly = true }
+<
+                  • line: line number, 0-based
+                  • start_col: start column, 0-based
+                  • end_col: start column, 0-based
+
+                • bufnr: buffer number
+                • priority: priority assigned to highlight (see below)
+                • namespace: the namespace managed by the semantic token
+                  highlighter. If you apply extmarks inside the rule, you
+                  likely want to use this namespace.
+      • {opts}  (table|nil) Optional keyword arguments:
+                • desc (string|nil): Description of rule, default "none"
+                • bufnr (number|nil): If specified, the rule only applies in
+                  the given buffer.
+                • ft (string|nil): if specified, applies to all files with the
+                  given filetype. Ignored if `bufnr` is specified. If neither
+                  is given, the rule applies globally.
+                • priority (number|nil): Use this priority for highlights
+                  returned by this rule, and supply it as the `priority`
+                  argument to `rule` callbacks. By default, this is 1 higher
+                  than the last rule added with `add_rule`, so rules added
+                  later have higher priority.
+                • pri_stack (number|nil): The number of priority level levels
```

Comment:
Can you explain more why this is needed? I'm not happy that this (only) affects internal state, which is very unusual through our API. Why isn't the explicit `priority` argument enough?

Reply:
I also feel that the `pri_stack` argument is awkward.

What it does is allow rules which return highlights and rules which set their own marks to coexist:
``` lua
st.add_rule(function(token)
  if token.modifiers.foo then
    return 'Foo'
  end
end)

st.add_rule(function(token, bufnr, priority, ns)
  local mark = function(hl_group, delta)
    vim.api.nvim_buf_set_extmark(bufnr, ns, token.line, token.start_col, {
      hl_group = hl_group,
      end_col = token.end_col,
      priority = priority + delta,
      strict = false,
    })
  end

  if token.modifier.defaultLibrary then mark('SemDefaultLibrary', 0) end
  if token.modifier.declaration then mark('SemDeclaration', 1) end
  if token.modifier.readOnly then mark('SemReadOnly', 2) end
end, { pri_stack = 3 } )

st.add_rule(function(token)
  if token.modifiers.bar then
    return 'Bar'
  end
end)
```

In that example, the rules happily stack up and work as expected. All marks from the middle rule are higher priority than the previous rule, and lower priority than the next rule. Without `pri_stack`, any rule that sets more than one priority level of marks is immediately banished to the realm of explicit priorities.

However,

- That particular rule could simply be rewritten as three normal rules, and I bet that will often be the case.
- Wanting to write rules that set extmarks and can also can be stacked with normal rules will probably always be an edge case, and certainly won't be immediately required.
- Keeping with the "add options as late as possible" philosophy, I'll remove it now.

Maybe there's a cleaner way to get the same ability?

Reply:
I think requiring explicitly setting priorities (for now) if someone cares about ordering would be best, yes. (I also would prefer not to set _implicit_ priorities based on order -- otherwise this is starting to get uncomfortably close to a new DSL. _If_ we go down that route of new runtime file type, we should design it properly. Nobody expects Lua functions to care about the order they're defined in!)